src/Rings/oscar_singular.jl

```diff
@@ -0,0 +1,446 @@
+    res += K(Int(c))*Ka^e[1]
+  end
+  return res
+end
+
+# fraction field of polynomial rings over QQ and Fp
+struct OscarSingularCoefficientRingMapFractionField{U, V, W, X} <: OscarSingularCoefficientRingMap{U, V}
+  R::U
+  S::V
+  g::W
+  Spoly::X #= we create univariate polynomials over S during the conversion =#
+end
+
+function _map_oscar_singular_univariate(Rx, g, f::Singular.spoly)
+  @assert base_ring(Rx) === domain(g)
+  @assert ngens(parent(f)) == 1
+  return Rx(preimage.(Ref(g), coefficients_of_univariate(f)))
+end
+
+function iso_oscar_singular_coeff_ring(F::AbstractAlgebra.Generic.FracField{<:PolyRingElem{T}}) where {T <: Union{FqFieldElem, QQFieldElem}}
+  R = base_ring(F)
+  g = iso_oscar_singular_coeff_ring(base_ring(R))
+  S, = Singular.FunctionField(codomain(g), [var(R)])
+  Sx, = polynomial_ring(S, :x; cached = false)
+  return OscarSingularCoefficientRingMapFractionField(F, S, g, Sx)
+end
+
+function preimage(f::OscarSingularCoefficientRingMapFractionField, a::Singular.n_transExt)
+  parent(a) !== codomain(f) && error("Element not in codomain")
+  F = domain(f)
+  R = base_ring(F)
+  n, d = Singular.n_transExt_to_spoly.([numerator(a), denominator(a)]; cached = false)
+  return F(_map_oscar_singular_univariate(R, f.g, n), _map_oscar_singular_univariate(R, f.g, d))
+end
+
+function image(f::OscarSingularCoefficientRingMapFractionField, a)
+  parent(a) !== domain(f) && error("Element not in domain")
+  F = base_ring(base_ring(domain(f))) # the F in domain(f) = F(X)
+  K = codomain(f)
+  @assert Singular.transcendence_degree(K) == 1 "wrong number of generators"
+  t, = Singular.transcendence_basis(K)
+  # It must be a transcendental extension of Q or Fp (other things are not supported)
+  n = map_coefficients(numerator(a); parent = f.Spoly) do x
+    if F isa FinField
+      K(lift(ZZ, x))
+    else
+      K(f.g(x))
+    end
+  end
+  d = map_coefficients(denominator(a)) do x
+    if F isa FinField
+      K(lift(ZZ, x))
+    else
+      K(f.g(x))
+    end
+  end
+  return divexact(n(t), d(t))
+end
+
+# rational function field
+struct OscarSingularCoefficientRingMapRationalFunctionField{U, V, W} <: OscarSingularCoefficientRingMap{U, V}
+  R::U
+  S::V
+  g::W
+end
+
+function iso_oscar_singular_coeff_ring(R::Generic.RationalFunctionField)
+  g = iso_oscar_singular_coeff_ring(R.fraction_field)
+  return OscarSingularCoefficientRingMapRationalFunctionField(R, codomain(g), g)
+end
+
+function image(f::OscarSingularCoefficientRingMapRationalFunctionField, a)
+  parent(a) !== domain(f) && error("Element not in domain")
+  return f.g(a.d)
+end
+
+function preimage(f::OscarSingularCoefficientRingMapRationalFunctionField, a)
+  parent(a) !== codomain(f) && error("Element not in codomain")
+  return domain(f)(preimage(f.g, a))
+end
+
+# catchall 
+struct OscarSingularCoefficientRingMapGeneric{U, V} <: OscarSingularCoefficientRingMap{U, V}
+  R::U
+  S::V
+end
+
```

Comment:
Now `OscarSingularCoefficientRingMapGeneric` is defined twice, so gotta get rid of this one
```suggestion
```